homeWork6/lesson1/Program.cs

```diff
@@ -0,0 +1,217 @@
+    public uint AccountNumber
+    {
+        get => _accountNumber;
+        set =>_accountNumber = value;//++_counter;//_accountNumber = String.Format("{0:00000}", ++_counter);
+    }
+
+    //accountType
+    public enum AccountType { Current, Estimated, Credit, Deposit };
+    private AccountType _accountType;
+    public AccountType CurrentAccountType
+    {
+        get => _accountType;
+        set => _accountType = value;
+    }
+
+    //balance
+    private double _balance;
+    public double Balance
+    {
+        get => _balance;
+        set => _balance = value;
+    }
+
+    private static uint _counter = 0;
+    private static uint Counter
+    {
+        get => _counter;
+        //set => _counter = value;
+    }
+
+    public bool GetMoney(double value)
+    {
+        //снятие со счета
+        double balance = _balance;
+        //
+        if (balance - value >= 0)
+        {
+            //деньги можно снять, изменяем баланс
+            _balance -= value;
+            return true;
+        }
+        return false;
+    }
+
+    public bool SetMoney(double value)
+    {
+        //положить на счет
+        if (value > 0)
+        {
+            _balance += value;
+            return true;
+        }
+        return false;
+    }
+
+    public bool TransferMoney(BankAccount bankAccount, double amount)//double value
+    {
+        //перевести деньги на счет,
+        //bankAccount - объект, у которого снимаются деньги
+        if (bankAccount != null)
+        {
+            if (amount > 0 && bankAccount.Balance - amount >= 0)
+            {
+                //на счете есть нужная сумма
+                bankAccount.Balance -= amount;
+                this.SetMoney(amount);              
+            }
+            return true;
+        }
+        return false;
+    }
+
+    public static bool operator ==(BankAccount account1, BankAccount account2)
+    {
+        if (ReferenceEquals(account1, account2)) 
+        {
+            return true;
+        }
+            
+        if (ReferenceEquals(account1, null)) 
+        {
+            return false;
+        }
+            
+        if (ReferenceEquals(account2, null))
+        {
+            return false;
+        }
+        return account1.Equals(account2);
+    }
+
+    public static bool operator !=(BankAccount account1, BankAccount account2) => !(account1 == account2);
+
+    public override bool Equals(object obj)
+    {
+        if (obj is not BankAccount account)
+        {
+           return false;
+        }
+        return AccountNumber == account.AccountNumber &&
+               Balance == account.Balance &&
+               CurrentAccountType == account.CurrentAccountType;
+    }
+
+    public override int GetHashCode()
+    {
+        int hashcode = _accountNumber.GetHashCode();
+        hashcode = 31 * hashcode + _accountType.GetHashCode();
+        hashcode = 31 * hashcode + _balance.GetHashCode();
+        return hashcode;

```

Comment:
Возможно переполнение!
```suggestion
        unchecked {
        int hashcode = _accountNumber.GetHashCode();
        hashcode = 31 * hashcode + _accountType.GetHashCode();
        hashcode = 31 * hashcode + _balance.GetHashCode();
        return hashcode;
       }
```